### PR TITLE
[gcs] Update all redis kv usage in python except function table

### PR DESCRIPTION
**Pull request by fishbone**
## Why are these changes needed?
This is part of redis removal project. In this PR all direct usage of redis got removed except function table.
Function table will be migrated in the next PR

## Related issue number
#19443

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by ericl — CHANGES_REQUESTED** [review]
Looks good, but could we namespace the KV keys by component? e.g. "ray.util.client.server:healthcheck", which makes it clear what component is doing the put. It would also be great if we make the keys constants everywhere.

**Comment by iycheng** [comment]
make namespace optional to make it backward compatible since it might be used by components outside of ray

File: dashboard/modules/reporter/reporter_head.py [hunk 1]
```diff
@@ -113,12 +114,16 @@ async def get_cluster_status(self, req):
...
         formatted_status_string = internal_kv._internal_kv_get(
-            DEBUG_AUTOSCALING_STATUS)
+            DEBUG_AUTOSCALING_STATUS,
+            namespace=ray_constants.KV_NAMESPACE_AUTOSCALER)
         formatted_status = json.loads(formatted_status_string.decode()
                                       ) if formatted_status_string else {}
-        error = internal_kv._internal_kv_get(DEBUG_AUTOSCALING_ERROR)
+        error = internal_kv._internal_kv_get(
+            DEBUG_AUTOSCALING_ERROR,
+            namespace=ray_constants.KV_NAMESPACE_AUTOSCALER)
```
Comment (ericl) [inline dashboard/modules/reporter/reporter_head.py:126]: Probably want to remove namespace for backwards compatibility of these for now.

File: python/ray/ray_constants.py [hunk 2]
```diff
@@ -278,3 +278,15 @@ def to_memory_units(memory_bytes, round_up):
 # The default gRPC max message size is 4 MiB, we use a larger number of 100 MiB
 # NOTE: This is equal to the C++ limit of (RAY_CONFIG::max_grpc_message_size)
 GRPC_CPP_MAX_MESSAGE_SIZE = 100 * 1024 * 1024
+
+# Internal kv namespaces
+KV_NAMESPACE_DASHBOARD = "dashboard"
+KV_NAMESPACE_SESSION = "session"
+KV_NAMESPACE_TRACING = "tracing"
+KV_NAMESPACE_PDB = "ray_pdb"
+# TODO(set autoscaler ns)
```
Comment (ericl) [inline python/ray/ray_constants.py:287]: Improve comment of why we set this to None for now?

File: python/ray/experimental/internal_kv.py [hunk 3]
```diff
@@ -25,35 +25,64 @@ def _initialize_internal_kv(gcs_client: GcsClient):
     _initialized = True
 
 
+__NS_START_CHAR = b"@"
```
Comment (ericl) [inline python/ray/experimental/internal_kv.py:28]: Comment.

**Review by ericl — COMMENTED** [review]
LGTM, some comments can be improved.

File: python/ray/experimental/internal_kv.py [hunk 4]
```diff
@@ -25,35 +25,64 @@ def _initialize_internal_kv(gcs_client: GcsClient):
...
+
+
 @client_mode_hook(auto_init=False)
 def _internal_kv_initialized():
     return global_gcs_client is not None
 
 
 @client_mode_hook(auto_init=False)
-def _internal_kv_get(key: Union[str, bytes]) -> bytes:
+def _internal_kv_get(key: Union[str, bytes],
```
Comment (mwtian) [inline python/ray/experimental/internal_kv.py:55]: I personally think it is better to have no leading underscore in the function name here, and one leading underscore for the make_key / parse_key functions. But it is ok to leave them as they are.

**Review by mwtian — APPROVED** [review]

File: python/ray/experimental/internal_kv.py (hunk 4 above)
Comment (iycheng) [reply python/ray/experimental/internal_kv.py:55]: it's hard to change the api right now. i'd prefer doing it later if we have time. the issue here is that, it's not just used by ray, but also other services. another reason is that ray client has proxy to this function.
this means we have to  1) keep both versions; 2) also make ray client use both versions. 3) update internal usage; 4) give time to let user update their usage.
so i'd prefer keeping it as now.

Comment (mwtian) [reply python/ray/experimental/internal_kv.py:55]: Sounds good.

**Comment by iycheng** [comment]
@ericl  could you check this one again?

**Comment by iycheng** [comment]
ping @ericl

**Review by ericl — APPROVED** [review]

**Review by architkulkarni — APPROVED** [review]
Looks good!  For the `test_runtime_env_working_dir.py`, were there any changes besides splitting it into two files?  Also, what was the reason for splitting it up? (Was it timing out in CI?)

File: dashboard/modules/snapshot/snapshot_head.py [hunk 5]
```diff
@@ -161,13 +160,17 @@ async def get_serve_info(self) -> Dict[str, Any]:
         # These are the keys we are interested in:
         # SERVE_CONTROLLER_NAME(+ optional random letters):SERVE_SNAPSHOT_KEY
 
-        serve_keys = _internal_kv_list(SERVE_CONTROLLER_NAME)
+        serve_keys = _internal_kv_list(
+            SERVE_CONTROLLER_NAME, namespace=ray_constants.KV_NAMESPACE_SERVE)
+        print("DBG:", serve_keys)
```
Comment (wuisawesome) [inline dashboard/modules/snapshot/snapshot_head.py:165]: Remove?

File: dashboard/tests/test_dashboard.py [hunk 6]
```diff
@@ -215,7 +222,7 @@ def test_http_get(enable_test_module, ray_start_with_dashboard):
             assert dump_info["result"] is True
             break
         except (AssertionError, requests.exceptions.ConnectionError) as e:
-            logger.info("Retry because of %s", e)
+            print("Retry because of %s", e)
```
Comment (wuisawesome) [inline dashboard/tests/test_dashboard.py:225]: revert?

**Comment by iycheng** [comment]
@architkulkarni this test actually is flaky. I checked some previous runs, and it says:

```
Stats over 2 runs: max = 915.0s, min = 792.6s, avg = 853.8s, dev = 61.2s

```
https://buildkite.com/ray-project/ray-builders-branch/builds/4571#e5f8efe2-c3e5-4fe0-8bc7-54d92eda6ef6

And the failure is due to timeout, so I split this one.

**Comment by architkulkarni** [comment]
I see, thanks for helping deflake it!

File: python/ray/ray_constants.py [hunk 7]
```diff
@@ -278,3 +278,22 @@ def to_memory_units(memory_bytes, round_up):
...
+KV_NAMESPACE_DASHBOARD = "dashboard"
+KV_NAMESPACE_SESSION = "session"
+KV_NAMESPACE_TRACING = "tracing"
+KV_NAMESPACE_PDB = "ray_pdb"
+
+# TODO: Set namespace for autoscaler
+# This will break backward compatibility between versions, where old version
+# autoscaler won't be able to scale up new version ray since the key will be
+# different with namespace.
+KV_NAMESPACE_AUTOSCALER = None
```
Comment (wuisawesome) [inline python/ray/ray_constants.py:292]: Can we move this constant into the autoscaler package? 

Also is there a migration plan? I feel like we should probably make this read an env var (that isn't set by default).

File: python/ray/tests/test_actor.py [hunk 8]
```diff
@@ -273,7 +272,7 @@ def __init__(self):
             pass
 
     Foo.remote()
-
+    # TODO: redis-removal kv
```
Comment (wuisawesome) [inline python/ray/tests/test_actor.py:275]: Just confirming my understanding, this is the definition of the actor that the function manager puts in redis right?

File: python/ray/tests/test_runtime_env_working_dir_2.py [hunk 9]
```diff
@@ -0,0 +1,410 @@
+from contextlib import contextmanager
```
Comment (wuisawesome) [inline python/ray/tests/test_runtime_env_working_dir_2.py:1]: Presumably this was just splitting the test into 2 files right?

**Review by wuisawesome — CHANGES_REQUESTED** [review]
1. I'm a little concerned about the dashboard changes since we're turning a bunch of async awaits into blocking operations. Do you think it's worth feature flagging?

2. Do we need a migration plan for the autoscaler namespace? 

3.  A few minor inline comments

File: python/ray/tests/test_runtime_env_working_dir_2.py (hunk 9 above)
Comment (iycheng) [reply python/ray/tests/test_runtime_env_working_dir_2.py:1]: yes

File: python/ray/tests/test_actor.py (hunk 8 above)
Comment (iycheng) [reply python/ray/tests/test_actor.py:275]: Yes, and it's not handled in this PR.

**Review by wuisawesome — APPROVED** [review]
Can you fix the autoscaler one before merging?

File: python/ray/ray_constants.py (hunk 7 above)
Comment (iycheng) [reply python/ray/ray_constants.py:292]: Hmm, I'm not sure about this. Any benefit of doing this?

**Comment by iycheng** [comment]
> 1. I'm a little concerned about the dashboard changes since we're turning a bunch of async awaits into blocking operations. Do you think it's worth feature flagging?
> 2. Do we need a migration plan for the autoscaler namespace?
> 3. A few minor inline comments

1. I think it's ok now since they are just kv get/put. I probably will add an async version in recent days.
2. The current plan is not doing anything for autoscaler. I'll move the namespace to autoscaler code.

**Pull request merged**